PVS-Studio static analyzer warning fixes (part 2)

dimhotepus opened the pull request:
Part 2 of fixes by PVS-Studio static analyzer tool.

dimhotepus committed:
[init] Initialize class members to safe defaults

dimhotepus committed:
[flow] Do not initialize var to same value twice

dimhotepus committed:
[flow] Remove zero side effects statement

dimhotepus committed:
[bounds] Do not overflow texture matrices array

dimhotepus committed:
[priority] Make reverse sqrt intention explicit

dimhotepus committed:
[perf] Implement IEquatable<T> to prevent boxing and reflection on comparisons

dimhotepus committed:
[perf] Use range for loop instead of iterators

dimhotepus committed:
[leak] Do not leak memory blocks

dimhotepus committed:
[alloc] Fix UB in deletion of T[] by delete

dimhotepus committed:
[bounds] Prevent out-of-bounds reads

* m_unweighted[16] accessed by c0 [0..16].
* m_unweighted[16] accessed by c0+c1 [0..16].

dimhotepus committed:
[macro] Close va_list after usage

dimhotepus committed:
[flow] Remove already true condition

dimhotepus committed:
[flow] Safe invocation of event subscribers

Event can be set to null after check, but before invoke. ?. prevents this.

dimhotepus committed:
[flow] Propagate null checks to prevent NullReferenceException

dimhotepus committed:
[flow] Add virtual destructor for type which has children and deleted

ide_base is base class, child is ide_impl. In void finalize(XRay::Editor::ide_base*& ide) we pass object via reference to pointer of base type and delete it.

So if ide_base has no virtual destructor, only its destructor is called, which lead to possible child resource leaks.

Xottab-DUTY commented on Externals/NVTT/src/nvimage/Filter.cpp:
```diff
@@ -329,7 +329,7 @@ Kernel2::Kernel2(const Kernel2 & k) : m_windowSize(k.m_windowSize)
 Kernel2::~Kernel2()
 {
-	delete m_data;
+	delete [] m_data;
```
```suggestion
	delete[] m_data;
```

Xottab-DUTY commented on Externals/NVTT/src/nvimage/Filter.cpp:
```diff
@@ -297,7 +297,7 @@ Kernel1::Kernel1(const Filter & f, int iscale, int samples/*= 32*/)
 Kernel1::~Kernel1()
 {
-	delete m_data;
+	delete [] m_data;
```
```suggestion
	delete[] m_data;
```

Xottab-DUTY commented on src/editors/xrPostprocessEditor/EditorEngine.cs:
```diff
@@ -6,17 +6,27 @@
+
+        public bool Equals(ColorMappingParams other)
+        {
+          return Influence == other.Influence && Texture == other.Texture;
```
```suggestion
            return Influence == other.Influence && Texture == other.Texture;
```

Xottab-DUTY commented on src/Layers/xrRender/DetailManager.cpp:
```diff
@@ -213,16 +213,19 @@ void CDetailManager::Unload()
+    for (vis_list& l : m_visibles)
+    {
+        l.clear();
+    }
```
Wastes too much space, introduces indentation to the right, too small – so, doesn't worth it.

dimhotepus replied:
Ok, restored.

dimhotepus committed:
[flow] Add virtual destructor for type which has children and deleted

property_holder is property_holder_base typedef, child is property_holder (not typedef). In ide_impl::destroy(property_holder*& property_holder) we pass object via reference to pointer of base type and delete it.

So if property_holder_base has no virtual destructor, only its destructor is called, which lead to possible child resource leaks.

dimhotepus committed:
[flow] Check result after dynamic_cast

We cast object to string and ignore result. If object is not string, we dereference nullptr later.

dimhotepus committed:
[leak] Do not throw potential OutOfMemoryException from destructor

Xottab-DUTY commented on src/utils/ETools/audio.cpp:
```diff
@@ -776,6 +792,7 @@ void setup_downmix(oe_enc_opt* opt)
     if (opt->channels != 2)
     {
         fprintf(stderr, "Internal error! Please report this bug.\n");
+        free(d);
```
It's better to just move the memory allocation down.

dimhotepus committed:
[flow] Deleting of gcroot m_container is nonsense

m_container is of gcroot type. Such type acts as a reference to the managed object or value type instance and is doing all the work when copying the object or value type instance.

gcroot has a destructor wich does all the job of cleaning resources for us.

dimhotepus committed:
[flow] Expected int type instead of bool one

Xottab-DUTY commented on src/utils/ETools/audio.cpp:
```diff
@@ -250,6 +252,7 @@ int aiff_open(FILE* in, oe_enc_opt* opt, unsigned char* buf, int buflen)
     if (fread(buffer, 1, len, in) < len)
     {
         fprintf(stderr, _("Warning: Unexpected EOF in reading AIFF header\n"));
+        free(aiff);
```
We can just move the `aiff` allocation down to the place where it's actually used.

`aiff->bigendian =` can be replaced with `int bigendian =`

dimhotepus replied:
Unfortunately, `aiff` is used in nested block, too. 

May be use `std::make_unique` + `release` where needed?

dimhotepus replied:
Nevermind, fixed by `int bigendian =`

Xottab-DUTY commented on src/utils/ETools/audio.cpp:
```diff
@@ -366,11 +375,15 @@ int wav_open(FILE* in, oe_enc_opt* opt, unsigned char* oldbuf, int buflen)
 
     if (!find_wav_chunk(in, "fmt ", &len))
+    {
+        free(wav);
```
We can just move the `wav` allocation down to the place where it's actually used.

Xottab-DUTY commented on src/Include/editor/ide.hpp:
```diff
@@ -22,6 +22,7 @@ class property_holder_holder;
 class ide_base
 {
 public:
+    virtual ~ide_base() {}
```
```suggestion
    virtual ~ide_base() = default;
```

Xottab-DUTY commented on src/Include/editor/property_holder_base.hpp:
```diff
@@ -126,6 +126,7 @@ class property_holder_base
     }; // enum value_editor_enum
 
 public:
+    virtual ~property_holder_base() {}
```
```suggestion
    virtual ~property_holder_base() = default;
```

dimhotepus committed:
[init] Initialize class members to safe defaults

Xottab-DUTY commented on src/editors/xrSdkControls/Controls/TreeViewSearchPanel/TreeViewSearchPanel.cs:
```diff
@@ -72,7 +72,7 @@ private void searchBox_KeyDown(object sender, KeyEventArgs e)
                     MessageBox.Show("There is no matches!");
                 else if (findedNodes.Count == 1)
                     resultLabel.Text = "Result: 1 / 1";
-                else if (findedNodes.Count > 1)
```
And now with have two `else { ... } else { ... }` branches

dimhotepus replied:
Sorry, output window has hidden last else from me. Just removed dead code.

dimhotepus committed:
[format] Fix UB for nonconsistent format and placeholder types

dimhotepus committed:
[format] Ensure format and type placeholders match

dimhotepus committed:
[style] Remove unneeded template param specification

dimhotepus committed:
[flow] No sense in passing references if args are not modified

dimhotepus committed:
[format] Fix format and placeholder type matching

Xottab-DUTY commented on src/utils/ETools/audio.cpp:
```diff
@@ -366,7 +366,9 @@ int wav_open(FILE* in, oe_enc_opt* opt, unsigned char* oldbuf, int buflen)
     if (!find_wav_chunk(in, "fmt ", &len))
+    {
         return 0; /* EOF */
+    }
```
Unneeded formatting

Xottab-DUTY commented on src/Layers/xrRender/DetailManager.cpp:
```diff
@@ -213,16 +213,17 @@ void CDetailManager::Unload()
     m_visibles[0].clear();
     m_visibles[1].clear();
     m_visibles[2].clear();
+
```
Unneeded formatting

Xottab-DUTY commented on src/utils/ETools/audio.cpp:
```diff
@@ -258,7 +257,7 @@ int aiff_open(FILE* in, oe_enc_opt* opt, unsigned char* buf, int buflen)
     format.rate = (int)read_IEEE80(buffer + 8);
 
-    aiff->bigendian = 1;
+    int bigendian = 1;
```
Was wrong a bit. Let's make it 100% correspond to `aifffile::bigendian` type
```suggestion
    short bigendian = 1;
```

Xottab-DUTY commented on src/Layers/xrRender/DetailManager.cpp:
```diff
@@ -213,16 +213,17 @@ void CDetailManager::Unload()
     hw_Unload();
 
-    for (DetailIt it = objects.begin(); it != objects.end(); it++)
+    for (CDetail* o : objects)
```
It's better to name it `detailObject`.

Xottab-DUTY commented on src/editors/xrWeatherEditor/property_converter_color.cpp:
```diff
@@ -83,7 +83,7 @@ bool property_converter_color::CanConvertFrom(ITypeDescriptorContext ^ context,
 Object ^ property_converter_color::ConvertFrom(ITypeDescriptorContext ^ context, CultureInfo ^ culture, Object ^ value)
 {
     String ^ string = dynamic_cast<String ^>(value);
-    if (!value)
```
I just checked the code and I see that this replacement is wrong and the code was written like that intentionally.

Xottab-DUTY replied:
But it makes sense to swap the code like that:
```
	if (!value)
		return (inherited::ConvertFrom(context, culture, value));

	String ^ string = dynamic_cast<String ^>(value);
	try
	{
		...
```

dimhotepus replied:
Good to catch it before the merge.

I'll move cast below value check than.

dimhotepus replied:
> But it makes sense to swap the code like that:
> 
> ```
> 	if (!value)
> 		return (inherited::ConvertFrom(context, culture, value));
> 
> 	String ^ string = dynamic_cast<String ^>(value);
> 	try
> 	{
> 		...
> ```

Done, thank you. Hope space doesn't matter here.

Xottab-DUTY replied:
Well, it doesn't look good, sorry
![oops](https://cdn.discordapp.com/emojis/639553094316851200.png?v=1)

dimhotepus replied:
It'll be better to use ClangFormat or similar tool to automate boring formatting task.

I'll fix formatting tomorrow, thanks!

dimhotepus replied:
Done!

Xottab-DUTY replied:
Thank you!

> It'll be better to use ClangFormat or similar tool to automate boring formatting task.

We have `.clang-format` file in the `src` dir. Doesn't it work?
